File: problems/group_anagrams/group_anagrams.py
import collections

from typing import List


def groupAnagrams(strs: List[str]) -> List[List[str]]:
    groups = collections.defaultdict(list)
    for s in strs:
        count = [0] * 26
        for c in s:
            count[ord(c) - ord("a")] += 1
        groups[tuple(count)].append(s)
    return list(groups.values())

File: problems/group_anagrams/group_anagrams.cc
#include "group_anagrams.h"

#include <algorithm>
#include <ranges>  // NOLINT(misc-include-cleaner) needed for std::ranges::sort
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

using namespace std;

vector<vector<string>> groupAnagrams(vector<string>& strs) {
    unordered_map<string, vector<string>> groups;
    groups.reserve(strs.size());

    for (const string& s : strs) {
        string key = s;
        std::ranges::sort(key);  // anagrams share the same sorted key
        groups[key].push_back(s);
    }

    vector<vector<string>> res;
    res.reserve(groups.size());
    for (auto& [k, v] : groups) {
        res.push_back(std::move(v));
    }
    return res;
}